### PR TITLE
change listPrefix in s3 file system to listobjectv2

### PR DESCRIPTION
**Pull request by zijian-qiao**
According to [AWS S3 document](https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html),
```
Important
    This section describes the latest revision of the API. 
    We recommend that you use this revised API for application development.
    For backward compatibility, Amazon S3 continues to support the prior version of this API, ListObjects.
```
And after some tests, V2 is indeed better. So we change ListObjects to ListObjectsV2 in PrestoS3FileSystem.

### CODE REVIEW

File: presto-hive/src/main/java/io/prestosql/plugin/hive/s3/PrestoS3FileSystem.java [hunk 1]
```diff
@@ -514,29 +514,37 @@ public boolean mkdirs(Path f, FsPermission permission)
...
             @Override
-            protected ObjectListing computeNext(ObjectListing previous)
+            protected ListObjectsV2Result computeNext(ListObjectsV2Result previous)
             {
                 if (!previous.isTruncated()) {
                     return null;
                 }
-                return s3.listNextBatchOfObjects(previous);
+
+                final ListObjectsV2Request nextRequest = new ListObjectsV2Request()
```
Comment (sopel39) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/s3/PrestoS3FileSystem.java]: you can use previous `request` and just set continuation token, see: https://docs.aws.amazon.com/AmazonS3/latest/dev/ListingObjectKeysUsingJava.html

File: presto-hive/src/test/java/io/prestosql/plugin/hive/s3/MockAmazonS3.java [hunk 2]
```diff
@@ -117,6 +119,28 @@ public ObjectListing listObjects(ListObjectsRequest listObjectsRequest)
         return listing;
     }
 
+    @Override
```
Comment (sopel39) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/s3/MockAmazonS3.java]: can we remove `listObjects` method?

**Comment by tooptoop4** [comment]
gentle ping @zijian-qiao

File: presto-hive/src/test/java/io/prestosql/plugin/hive/s3/MockAmazonS3.java (hunk 2 above)
Comment (zijian-qiao) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/s3/MockAmazonS3.java]: I think it should be ok. I will try.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/s3/PrestoS3FileSystem.java (hunk 1 above)
Comment (zijian-qiao) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/s3/PrestoS3FileSystem.java]: Sorry for the late reply, I will change code and make a basic test today.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/s3/MockAmazonS3.java [hunk 3]
```diff
@@ -96,25 +96,25 @@ public PutObjectResult putObject(PutObjectRequest putObjectRequest)
...
     {
-        ObjectListing listing = new ObjectListing();
+        ListObjectsV2Result listingV2 = new ListObjectsV2Result();
 
         S3ObjectSummary standard = new S3ObjectSummary();
         standard.setStorageClass(StorageClass.Standard.toString());
         standard.setKey("test/standard");
         standard.setLastModified(new Date());
-        listing.getObjectSummaries().add(standard);
+        listingV2.getObjectSummaries().add(standard);
```
Comment (sopel39) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/s3/MockAmazonS3.java]: Could we also have test that uses `setContinuationToken` path?

**Comment by findepi** [comment]
I checked with MinIO team and they support `listObjectV2`.

**Comment by sopel39** [comment]
So it looks like it's good to go.
@zijian-qiao could you additionally run `io.prestosql.plugin.hive.TestHiveFileSystemS3` tests?

**Comment by tooptoop4** [comment]
gentle ping @zijian-qiao

File: presto-hive/src/test/java/io/prestosql/plugin/hive/s3/MockAmazonS3.java (hunk 3 above)
Comment (zijian-qiao) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/s3/MockAmazonS3.java]: Added. But I'm not sure if it is a right way. Please help to review, thanks~

**Comment by zijian-qiao** [comment]
> So it looks like it's good to go.
> @zijian-qiao could you additionally run `io.prestosql.plugin.hive.TestHiveFileSystemS3` tests?

Sorry, I don't know how to run this test locally with correct parameters. Or should I initialize the test context myself?

**Comment by sopel39** [comment]
> Sorry, I don't know how to run this test locally with correct parameters.

There is `presto-hive-hadoop2/bin/run_hive_s3_tests.sh`, but you need to have a test S3 bucket for this. Do you have such bucket?

**Comment by zijian-qiao** [comment]
> There is `presto-hive-hadoop2/bin/run_hive_s3_tests.sh`, but you need to have a test S3 bucket for this. Do you have such bucket?

Yes, I have, I will try.

**Comment by sopel39** [comment]
Merged, thanks!

**Comment by zijian-qiao** [comment]
> Merged, thanks!

Thanks!

**Pull request merged**